ddl/foreign_key_test.go

```diff
@@ -15,181 +15,572 @@
+			create: "create table t2 (a int, b varchar(10), foreign key fk_b(b) references t1(a));",
+			err:    "[schema:1822]Failed to add the foreign key constraint. Missing index for constraint 'fk_b' in the referenced table 't1'",
+		},
+		{
+			refer:  "create table t1 (id int key, a int, index(a));",
+			create: "create table t2 (a int, b int, foreign key fk_b(b) references t1(id, a));",
+			err:    "[schema:1239]Incorrect foreign key definition for 'fk_b': Key reference and table reference don't match",
+		},
+		{
+			create: "create table t2 (a int key, foreign key (a) references t2(a));",
+			err:    "[schema:1215]Cannot add foreign key constraint",
+		},
+		{
+			create: "create table t2 (a int, b int, index(a,b), index(b,a), foreign key (a,b) references t2(a,b));",
+			err:    "[schema:1215]Cannot add foreign key constraint",
+		},
+		{
+			create: "create table t2 (a int, b int, index(a,b), foreign key (a,b) references t2(b,a));",
+			err:    "[schema:1822]Failed to add the foreign key constraint. Missing index for constraint 'fk_1' in the referenced table 't2'",
+		},
+		{
+			prepare: []string{
+				"set @@foreign_key_checks=0;",
+				"create table t2 (a int, b int, index(a), foreign key (a) references t1(id));",
+			},
+			create: "create table t1 (id int, a int);",
+			err:    "[schema:1822]Failed to add the foreign key constraint. Missing index for constraint 'fk_1' in the referenced table 't1'",
+		},
+		{
+			prepare: []string{
+				"set @@foreign_key_checks=0;",
+				"create table t2 (a int, b int, index(a), foreign key (a) references t1(id));",
+			},
+			create: "create table t1 (id bigint key, a int);",
+			err:    "[ddl:3780]Referencing column 'a' and referenced column 'id' in foreign key constraint 'fk_1' are incompatible.",
+		},
+		{
+			// foreign key is not support in temporary table.
+			refer:  "create temporary table t1 (id int key, b int, index(b))",
+			create: "create table t2 (a int, b int, foreign key fk(b) references t1(b))",
+			err:    "[schema:1824]Failed to open the referenced table 't1'",
+		},
+		{
+			// foreign key is not support in temporary table.
+			refer:  "create global temporary table t1 (id int key, b int, index(b)) on commit delete rows",
+			create: "create table t2 (a int, b int, foreign key fk(b) references t1(b))",
+			err:    "[schema:1215]Cannot add foreign key constraint",
+		},
+		{
+			// foreign key is not support in temporary table.
+			refer:  "create table t1 (id int key, b int, index(b))",
+			create: "create temporary table t2 (a int, b int, foreign key fk(b) references t1(b))",
+			err:    "[schema:1215]Cannot add foreign key constraint",
+		},
+		{
+			// foreign key is not support in temporary table.
+			refer:  "create table t1 (id int key, b int, index(b))",
+			create: "create global temporary table t2 (a int, b int, foreign key fk(b) references t1(b)) on commit delete rows",
+			err:    "[schema:1215]Cannot add foreign key constraint",
+		},
+	}
```

Comment:
Please add the following test cases:
```sql
create table t (a int, foreign key ``(a) references t1(a));
create table t (a int, constraint `` foreign key (a) references t1(a));
```

The expected error is `ERROR 1280 (42000): Incorrect index name ''`.

Reply:
great, done.